src/platform/backends/qemu/qemu_virtual_machine.cpp

```diff
@@ -194,12 +198,32 @@ void mp::QemuVirtualMachine::on_shutdown()
 
 std::string mp::QemuVirtualMachine::ssh_hostname()
 {
-    return ip.as_string();
+    return ipv4();
 }
 
 std::string mp::QemuVirtualMachine::ipv4()
 {
-    return ip.as_string();
+    if (*ip == mp::IPAddress{0xFFFFFFFF})
+    {
+        using namespace std::literals::chrono_literals;
+
+        auto deadline = std::chrono::steady_clock::now() + std::chrono::minutes(2);
+
+        while (std::chrono::steady_clock::now() < deadline)
+        {
+            QCoreApplication::processEvents();
+            ip = std::make_unique<mp::IPAddress>(dnsmasq_server->get_ip_for(mac_addr));
+
+            if (*ip != mp::IPAddress{0xFFFFFFFF})
```

Comment:
With ip as a std::experimental::optional, I believe this can become if(ip) return ip.value().as_string();